src/s2/s2region_term_indexer.cc

```diff
@@ -194,29 +215,34 @@ vector<string> S2RegionTermIndexer::GetIndexTermsForCanonicalCovering(
           prev_id.parent(level) == ancestor_id) {
         break;  // We have already processed this cell and its ancestors.
       }
-      terms.push_back(GetTerm(TermType::ANCESTOR, ancestor_id, prefix));
+      terms->push_back(GetTerm(TermType::ANCESTOR, ancestor_id, prefix));
     }
     prev_id = id;
   }
-  return terms;
 }
 
 vector<string> S2RegionTermIndexer::GetQueryTerms(const S2Point& point,
                                                   string_view prefix) {
+  vector<string> terms;
+  GetQueryTerms(point, prefix, &terms);
+  return terms;
+}
+
+void S2RegionTermIndexer::GetQueryTerms(const S2Point& point,
+                                        string_view prefix,
+                                        vector<string>* terms) {
```

Comment:
Please split changing the type signature into a separate PR. We can discuss it there. Does this have a measurable and meaningful performance difference for you? I would have guessed that computing the covering takes way longer than reallocating vectors.

https://google.github.io/eng-practices/review/developer/small-cls.html

Reply:
I don't want it, it's conflicting changes. It's not convenient to make separate PRs and then resolve conflict in one to another

Reply:
Sure, it's better, but is it 10% better or 0.001% better? We need to weigh the complexity of the change vs. the performance improvements.  How much does this improve 1. time to build the whole index (or insert an object) 2. query time? (or some other user-observable action)

Reply:
> We need to weigh the complexity of the change vs. the performance improvements. 

But it's not affect users who doesn't want to use it

Why is it issue?
In general same api exists for S2RegionCoverer, simple and advanced

Why don't you want same for S2RegionCoverer?
https://github.com/google/s2geometry/blob/master/src/s2/s2region_coverer.h#L189

Could you explain please?

Why for coverer it's ok, but for indexer it's too difficult? They usage are pretty same

> Sure, it's better, but is it 10% better or 0.001% better? 

I don't measure this part, but it depends on what we do.

Reply:
> Why don't you want same as for S2RegionCoverer?

The reason there are two here is more historical than technical. The `vector*` version was there since ~2005, and the version returning `S2CellUnion` was added in 2017. In that time, the style guide changed from recommending pointer args for output parameters to recommending return values.

https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

The reason it changed is that the performance difference is usually not big .

> I don't measure this part, but it depends on what we do.

Making performance optimizations without measurement is likely to misdirect effort.

If this were written from scratch, there would probably be no `vector*` version, or at least not until benchmarks demonstrated the utility.

Reply:
I really don't understand what the problem I want to make single allocation for N point, now it's impossible, after this commit possible, it's extension.

It's not difficult or something...
It's same as other classes api (if it's only historical reason deprecate it! I don't think so, vector is more optimal)

Reply:
https://github.com/google/s2geometry/pull/304 Done (I don't want to merge this branch, just push it for add ability to reproduce)

Reply:
Same but maybe a little lower for covering if we already has it.

(Thats reason why I add new overload only for FromCanonical, because compute covering is expensive, and in that function it's unnecessary)

Reply:
If I be honest I think it's possible to make benchmark better for old api. (in such case this benchmark became useless, because it will be reuse same memory (one in vector, another in allocator) in both overloads).

For points is possible to call exact vector::reserve, but I don't think it's good change because for regions it's impossible and it also makes new overload a little worse (unnecessary reserve)

And new overload make possible tuning on user side if necessary (if it's not necessary just use old)

Reply:
> Measure any changes is impossible commonly we measure changes about what we cannot proof logical.

I'm not sure what you mean here. Why is measurement impossible?

> For an example using hashmap instead of linear search for more than few elements
> I don't think for such changes needs benchmarks

hashmap vs linear search are generally the same code complexity. `vector*` or `vector&` is more complex because now you have to document what happens to the existing state, have to clear it somewhere, etc.

> Has s2 any microbenchmarks? If it's too critical for you I can add another one for show the difference between this overloads

Yes, but not for these functions.

Microbenchmarks are a means, not an end. If you profiled your whole system, and found a meaningful amount of time spent in `S2RegionTermIndexer::Get*Terms`, then microbenchmarks could be used to direct optimizations.

Reply:
> Why is measurement impossible?

measurement **all** impossible.
Because it costs time, and I have other tasks with more priority

Reply:
> In that time, the style guide changed from recommending pointer args for output parameters to recommending return values

But it's not about style, it's different behavior.

I think (maybe I'm not right, so please correct me) style guide more about
`absl::Status DoSomething(...);` vs `bool/void DoSomething(..., absl::Status*);`

Reply:
No, this case is covered, too. https://abseil.io/tips/176

Do the simpler thing unless there is *meaningful* performance difference.

Reply:
Firstly it's tips not rules. Also it's about simple and advanced overloads, it's not like only one. So if developer have questions about it it can just don't use it, and use it only if really needed.
Secondly it's meaningful for me.

At least for points it's easy measurable change.
Of course you can suggest reserve, but in such case needs idea how to force allocator to don't reuse same memory in benchmark (because it won't be on practice)

Reply:
> Secondly it's meaningful for me.

Ok, what are the numbers? How many terms do you have in average/worst case? How much time is it saving you? What is the larger operation you're trying to optimize and how long does that take?

Reply:
Depends on user. For one case nothing another more

Reply:
https://github.com/google/s2geometry/pull/303#discussion_r1066830083
https://github.com/google/s2geometry/pull/303#discussion_r1066833245

Waiting answer on other part of messages

Reply:
What exactly numbers do I need to provide to convince you?
It's look not constructive now

You ignore all arguments about it's not only stylistic difference.

Or is it impossible?
And I need to maintain patch/fork?

Reply:
> [#303 (comment)](https://github.com/google/s2geometry/pull/303#discussion_r1066830083) [#303 (comment)](https://github.com/google/s2geometry/pull/303#discussion_r1066833245)
> 
> Waiting answer on other part of messages

I don't see any questions there. What questions are you waiting for answers to?

> Depends on user. For one case nothing another more

Well, what's the range?

Reply:
> Well, what's the range?

from 0 to thousands of inserts in each thread.
You see in benchmark difference like 20-30%, of course it's not all job, but I think even 1% for just call other API is good enough.

> I don't see any questions there. What questions are you waiting for answers to?

You sent a tip, in this tip have two caveats exactly about this situation, and also it have example with pretty same issue.

So I waiting why do you think this tips is  good for your arguments not for mine.

Reply:
> You ignore all arguments about it's not only stylistic difference.

No, I agree it's more than style. We need to trade style against performance, so explain the performance impact.

> What exactly numbers do I need to provide to convince you?

Most convincing would be an arangodb load test where this change decreases latency by X%/T ms.

But you can start with how many index/query terms you're dealing with. Maybe the covering options are producing too many terms, and there's some other problem that should be solved.

Reply:
> We need to trade style against performance, so explain the performance impact.

It's not trade. Old overloads still exist!

Reply:
> But you can start with how many index/query terms you're dealing with. Maybe the covering options are producing too many terms, and there's some other problem that should be solved.

For an example I wrote count of inserts.
We have points only index, which can be created with default indexer options.

So one of example is exactly like in benchmark.

Reply:
> It's not trade.

It's extension

Reply:
https://github.com/abseil/abseil-cpp/blob/master/absl/strings/str_format.h

For an example it has two versions:
void StrAppendFormat and string StrFormat

Reply:
> We have points only index, which can be created with default indexer options.

This part of the change is good.

> It's extension

You want to make a performance change. Explain the performance impact in a time unit.

Reply:
> You want to make a performance change. Explain the performance impact in a time unit.

Already done

Reply:
> This part of the change is good

I mean we have **index** with points only. Every allocation cost at least 10ns (on practice more), so if we make 1 instead of N it decrease overall time at least on N*10ns

Query with points only it's other part of PR.

Reply:
> Every allocation cost at least 10ns (on practice more), so if we make 1 instead of N it decrease overall time at least on (N-1)*10ns

It's significant for me not because it's big but because it costs nothing to me, I just should use other API
Even small improvements still good, if don't make code significantly more difficult

Reply:
Concrete for me for query it will be same.
I can only see the difference only for indexing, but I thought it's better to have a symmetrical API.

Maybe as a compromise I can keep the new indexing API but remove the new querying API (if you think that would be better)